File: nodejs/test/conformance.test.js
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { describe, it } from 'node:test';

import {
  FORMAT_OPTION_MORE_VERBOSE,
  FORMAT_OPTION_NORMAL,
  FORMAT_OPTION_SIMPLE,
  FORMAT_OPTION_SIMPLEST,
  FORMAT_OPTION_VERBOSE,
  PreferredQuote,
  QuoteStrategy,
  formatType,
  formatTypeVerboseAnnotationOmit,
  formatTypeVerboseAnnotationPrimary,
  formatValue,
  literalFormatConfig,
  simpleFormatConfig,
  spannerCliFormatConfig,
} from '../lib/index.js';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const CONFORMANCE_PATH = path.resolve(__dirname, '../../testdata/conformance.json');

const TYPE_PRESETS = {
  simplest: FORMAT_OPTION_SIMPLEST,
  simple: FORMAT_OPTION_SIMPLE,
  normal: FORMAT_OPTION_NORMAL,
  verbose: FORMAT_OPTION_VERBOSE,
  more_verbose: FORMAT_OPTION_MORE_VERBOSE,
};

const QUOTE_POLICIES = {
  legacy_double: { strategy: QuoteStrategy.LEGACY, preferredQuote: PreferredQuote.DOUBLE },
  legacy_single: { strategy: QuoteStrategy.LEGACY, preferredQuote: PreferredQuote.SINGLE },
  always_double: { strategy: QuoteStrategy.ALWAYS, preferredQuote: PreferredQuote.DOUBLE },
  always_single: { strategy: QuoteStrategy.ALWAYS, preferredQuote: PreferredQuote.SINGLE },
  min_escape_double: { strategy: QuoteStrategy.MIN_ESCAPE, preferredQuote: PreferredQuote.DOUBLE },
  min_escape_single: { strategy: QuoteStrategy.MIN_ESCAPE, preferredQuote: PreferredQuote.SINGLE },
};

/** @returns {import('../lib/index.js')} */
function loadConformance() {
  return JSON.parse(readFileSync(CONFORMANCE_PATH, 'utf8'));
}

/**
 * @param {string} name
 * @param {Record<string, unknown>} typ
 */
function formatTypePreset(name, typ) {
  if (name === 'verbose_annotation_omit') {
    return formatTypeVerboseAnnotationOmit(typ);
  }
  if (name === 'verbose_annotation_primary') {
    return formatTypeVerboseAnnotationPrimary(typ);
  }
  return formatType(typ, TYPE_PRESETS[name]);
}

const conformance = loadConformance();

describe('type conformance', () => {
  for (const preset of [
    ...Object.keys(TYPE_PRESETS),
    'verbose_annotation_omit',
    'verbose_annotation_primary',
  ]) {
    it(`preset ${preset}`, () => {
      for (const testCase of conformance.type_cases) {
        const got = formatTypePreset(preset, testCase.type);
        const want = testCase.expected[preset];
        assert.equal(
          got,
          want,
          `type case ${testCase.name} preset ${preset}: got ${JSON.stringify(got)} want ${JSON.stringify(want)}`,
        );
      }
    });
  }
});

describe('value conformance', () => {
  for (const preset of ['simple', 'literal', 'spanner_cli']) {
    it(`preset ${preset}`, () => {
      const config = preset === 'simple'
        ? simpleFormatConfig()
        : preset === 'literal'
          ? literalFormatConfig()
          : spannerCliFormatConfig();

      for (const testCase of conformance.value_cases) {
        const got = formatValue(testCase.type, testCase.value, config);
        const want = testCase.expected[preset];
        assert.equal(
          got,
          want,
          `value case ${testCase.name} preset ${preset}: got ${JSON.stringify(got)} want ${JSON.stringify(want)}`,
        );
      }
    });
  }
});

describe('literal quote conformance', () => {
  for (const policyName of Object.keys(QUOTE_POLICIES)) {
    it(`policy ${policyName}`, () => {
      const config = literalFormatConfig(QUOTE_POLICIES[policyName]);
      for (const testCase of conformance.value_cases) {
        const got = formatValue(testCase.type, testCase.value, config);
        const want = testCase.expected.literal_quotes[policyName];
        assert.equal(
          got,
          want,
          `value case ${testCase.name} quote ${policyName}: got ${JSON.stringify(got)} want ${JSON.stringify(want)}`,
        );
      }
    });
  }
});

File: cpp/test/conformance_test.cpp
#define CATCH_CONFIG_MAIN
#include <catch2/catch.hpp>

#include <fstream>
#include <string>
#include <unordered_map>

#include <nlohmann/json.hpp>

#include "spanvalue/spanvalue.hpp"

namespace {

using Json = nlohmann::json;

Json load_conformance(const std::string& path) {
  std::ifstream in(path);
  REQUIRE(in.is_open());
  Json data;
  in >> data;
  return data;
}

}  // namespace

#ifndef SPANVALUE_CONFORMANCE_PATH
#define SPANVALUE_CONFORMANCE_PATH "../testdata/conformance.json"
#endif

TEST_CASE("type conformance cases", "[conformance][type]") {
  const Json conformance = load_conformance(SPANVALUE_CONFORMANCE_PATH);

  const std::unordered_map<std::string, spanvalue::FormatOption> presets = {
      {"simplest", spanvalue::format_option_simplest()},
      {"simple", spanvalue::format_option_simple()},
      {"normal", spanvalue::format_option_normal()},
      {"verbose", spanvalue::format_option_verbose()},
      {"more_verbose", spanvalue::format_option_more_verbose()},
  };

  for (const auto& [preset_name, option] : presets) {
    for (const Json& case_obj : conformance.at("type_cases")) {
      const std::string got = spanvalue::format_type(case_obj.at("type"), option);
      const std::string want = case_obj.at("expected").at(preset_name);
      INFO("type case " << case_obj.at("name").get<std::string>() << " preset " << preset_name);
      REQUIRE(got == want);
    }
  }

  for (const Json& case_obj : conformance.at("type_cases")) {
    const std::string got = spanvalue::format_type_verbose_annotation_omit(case_obj.at("type"));
    const std::string want = case_obj.at("expected").at("verbose_annotation_omit");
    INFO("type case " << case_obj.at("name").get<std::string>() << " preset verbose_annotation_omit");
    REQUIRE(got == want);
  }

  for (const Json& case_obj : conformance.at("type_cases")) {
    const std::string got = spanvalue::format_type_verbose_annotation_primary(case_obj.at("type"));
    const std::string want = case_obj.at("expected").at("verbose_annotation_primary");
    INFO("type case " << case_obj.at("name").get<std::string>() << " preset verbose_annotation_primary");
    REQUIRE(got == want);
  }
}

TEST_CASE("value conformance cases", "[conformance][value]") {
  const Json conformance = load_conformance(SPANVALUE_CONFORMANCE_PATH);

  const std::unordered_map<std::string, spanvalue::FormatConfig> configs = {
      {"simple", spanvalue::simple_format_config()},
      {"literal", spanvalue::literal_format_config()},
      {"spanner_cli", spanvalue::spanner_cli_format_config()},
  };

  for (const auto& [preset_name, config] : configs) {
    for (const Json& case_obj : conformance.at("value_cases")) {
      const std::string got =
          spanvalue::format_value(case_obj.at("type"), case_obj.at("value"), config);
      const std::string want = case_obj.at("expected").at(preset_name);
      INFO("value case " << case_obj.at("name").get<std::string>() << " preset " << preset_name);
      REQUIRE(got == want);
    }
  }
}

TEST_CASE("value literal quote policies", "[conformance][quotes]") {
  const Json conformance = load_conformance(SPANVALUE_CONFORMANCE_PATH);

  const std::unordered_map<std::string, spanvalue::LiteralQuoteConfig> policies = {
      {"legacy_double",
       spanvalue::LiteralQuoteConfig{spanvalue::QuoteStrategy::kLegacy,
                                     spanvalue::PreferredQuote::kDouble}},
      {"legacy_single",
       spanvalue::LiteralQuoteConfig{spanvalue::QuoteStrategy::kLegacy,
                                     spanvalue::PreferredQuote::kSingle}},
      {"always_double",
       spanvalue::LiteralQuoteConfig{spanvalue::QuoteStrategy::kAlways,
                                     spanvalue::PreferredQuote::kDouble}},
      {"always_single",
       spanvalue::LiteralQuoteConfig{spanvalue::QuoteStrategy::kAlways,
                                     spanvalue::PreferredQuote::kSingle}},
      {"min_escape_double",
       spanvalue::LiteralQuoteConfig{spanvalue::QuoteStrategy::kMinEscape,
                                     spanvalue::PreferredQuote::kDouble}},
      {"min_escape_single",
       spanvalue::LiteralQuoteConfig{spanvalue::QuoteStrategy::kMinEscape,
                                     spanvalue::PreferredQuote::kSingle}},
  };

  for (const auto& [policy_name, quote] : policies) {
    const spanvalue::FormatConfig config = spanvalue::literal_format_config(quote);
    for (const Json& case_obj : conformance.at("value_cases")) {
      const std::string got =
          spanvalue::format_value(case_obj.at("type"), case_obj.at("value"), config);
      const std::string want = case_obj.at("expected").at("literal_quotes").at(policy_name);
      INFO("value case " << case_obj.at("name").get<std::string>() << " quote " << policy_name);
      REQUIRE(got == want);
    }
  }
}
